rfc-144-consistent-cache-expiry.md

```diff
@@ -0,0 +1,186 @@
+There can be a 15 minute delay between publishing the banner and the banner
+appearing on all www.gov.uk pages.
+
+Proposed in this RFC:
+
+1. we will reduce TTLs on pages to a consistent 5 minutes
+1. we will not invalidate caches when updating the emergency banner, instead we will wait for 10 minutes
+
+[emergency banner]: https://docs.publishing.service.gov.uk/manual/emergency-publishing.html#header
+
+## Context
+
+For greater reliability GOV.UK runs several caches.
+
+At the edge, we run two main caches:
+
+- Fastly (CDN) caches all www.gov.uk pages, which honours the [Cache-Control]
+  header provided by Origin applications.
+- The [mirror cache] crawls www.gov.uk and stores all pages in multiple S3 buckets
+  and Google GCS for failover purposes when an Origin service is down and the
+  Fastly grace period for cached content has ended.
+
+On our Origin servers, we make use of several caches:
+
+- Varnish (self-hosted on cache machines) also caches all www.gov.uk pages
+  served by Frontend apps.
+- Frontend apps use the Rails cache (backed by Memcached, Redis, or File cache)
+  to [cache templates] served by Static for [60 seconds][slimmer-cache].
+- Static templates are [cached][static-cache] a fourth time by the Static
+  application, to enable Nginx to serve the templates.
+
+## Problem
+
+The emergency banner deploy process is unnecessarily fragile and hard to debug
+due to the "display banner instantly" feature.
+
+The "display banner instantly" feature is implemented using cache invalidation.
+After updating the Static templates to include the emergency banner text,
+the Jenkins job does an in-order purge of four layers of caches in which the
+Static templates are stored.
+
+The feature is necessary only because www.gov.uk pages generally set a cache TTL
+of 30 minutes and we have assumed that emergency banners need to be displayed instantly.
+
+The "display banner instantly" feature is not required, and when our
+implementation (cache invalidation) fails it has been hard to debug.
+
+[mirror cache]: https://docs.publishing.service.gov.uk/manual/fall-back-to-mirror.html
+[Cache-Control]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#the_cache-control_header
+[cache templates]: https://github.com/alphagov/slimmer#caching
+[slimmer-cache]: https://github.com/alphagov/slimmer/blob/b66fcbe9b667b2d946c5922ef42e619a6366c278/lib/slimmer.rb#L10
+[static-cache]: https://github.com/alphagov/static/blob/main/app/controllers/root_controller.rb#L8
+
+## Proposal
+
+The following commitments aim to make the emergency banner deploy process more
+reliable:
+
+* We will remove the "display banner instantly" feature from the
+  Deploy Emergency Banner job.
+  This will involve removing the four cache invalidation sub-jobs.
+* We will ensure that the emergency banner is displayed within 15 minutes.
+  This will involve reducing the default caching time for GOV.UK pages from 30m to 5m.
```

Comment:
Not in scope of this RFC, but do you think this means we could possibly get rid of the cache clearing service in the future?

Reply:
Ah yes, you mention this at the end